crates/matrix-sdk/src/room/timeline/mod.rs

```diff
@@ -88,14 +85,26 @@ impl Timeline {
         Self::from_inner(Arc::new(TimelineInner::new(room.to_owned())), None)
     }
 
+    #[cfg(feature = "experimental-sliding-sync")]
     pub(crate) async fn with_events(
         room: &room::Common,
         prev_token: Option<String>,
-        events: Vec<SyncTimelineEvent>,
+        events: Vec<matrix_sdk_base::deserialized_responses::SyncTimelineEvent>,
     ) -> Self {
         let mut inner = TimelineInner::new(room.to_owned());
         inner.add_initial_events(events).await;
-        Self::from_inner(Arc::new(inner), prev_token)
+
+        // The events we're injecting might be encrypted events, but we might
+        // have received the room key to decrypt them while nobody was listening to the
+        // `m.room_key` event, let's retry now.
+        //
+        // TODO: We could spawn a task here and put this into the background, though it
+        // might not be worth it depending on the number of events we injected.
+        // Some measuring needs to be done.
+        let timeline = Self::from_inner(Arc::new(inner), prev_token);
```

Comment:
I think it would be clearer if this comment came after the `Self::from_inner` call, right before `retry_decryption_for_all_events`.